File: languages/cpp/assessments/03-advanced/solutions/main.cpp
#include <algorithm>
#include <numeric>
#include <cmath>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

using namespace std;

class Shape {
  public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
    virtual string name() const = 0;
};

class Circle : public Shape {
  public:
    explicit Circle(double radiusValue) : radius(radiusValue) {}

    double area() const override { return 3.141592653589793 * radius * radius; }

    string name() const override { return "Circle"; }

  private:
    double radius;
};

class Rectangle : public Shape {
  public:
    Rectangle(double widthValue, double heightValue) : width(widthValue), height(heightValue) {}

    double area() const override { return width * height; }

    string name() const override { return "Rectangle"; }

  private:
    double width;
    double height;
};

template <typename T> void printVector(const vector<T>& values, const string& label) {
    cout << label << ": [";
    for (size_t i = 0; i < values.size(); ++i) {
        cout << values[i];
        if (i + 1 < values.size()) {
            cout << ", ";
        }
    }
    cout << "]\n";
}

int main() {
    vector<unique_ptr<Shape>> shapes;
    shapes.push_back(make_unique<Circle>(2.0));
    shapes.push_back(make_unique<Rectangle>(3.0, 4.0));
    shapes.push_back(make_unique<Circle>(1.5));

    vector<double> areas;
    areas.reserve(shapes.size());

    cout << "Shape areas:\n";
    for (const unique_ptr<Shape>& shape : shapes) {
        const double value = shape->area();
        areas.push_back(value);
        cout << "- " << shape->name() << ": " << value << '\n';
    }

    const double totalArea = accumulate(areas.begin(), areas.end(), 0.0);
    const double minArea = *min_element(areas.begin(), areas.end());
    const double maxArea = *max_element(areas.begin(), areas.end());

    cout << "Total area: " << totalArea << '\n';
    cout << "Minimum area: " << minArea << '\n';
    cout << "Maximum area: " << maxArea << '\n';

    vector<int> sampleCounts = {1, 2, 3, 4};
    printVector(sampleCounts, "Sample counts");
    printVector(areas, "Computed areas");

    return 0;
}

File: languages/csharp/assessments/03-advanced/solutions/main.cs
using System;
using System.Collections.Generic;

abstract class Shape
{
    public abstract double Area();
    public abstract string Name();
}

sealed class Circle : Shape
{
    private readonly double radius;

    public Circle(double radiusValue)
    {
        radius = radiusValue;
    }

    public override double Area()
    {
        return Math.PI * radius * radius;
    }

    public override string Name()
    {
        return "Circle";
    }
}

sealed class Rectangle : Shape
{
    private readonly double width;
    private readonly double height;

    public Rectangle(double widthValue, double heightValue)
    {
        width = widthValue;
        height = heightValue;
    }

    public override double Area()
    {
        return width * height;
    }

    public override string Name()
    {
        return "Rectangle";
    }
}

class Program
{
    static void PrintList<T>(List<T> values, string label)
    {
        Console.Write(label + ": [");
        for (int index = 0; index < values.Count; index++)
        {
            Console.Write(values[index]);
            if (index + 1 < values.Count)
            {
                Console.Write(", ");
            }
        }

        Console.WriteLine("]");
    }

    static void Main()
    {
        List<Shape> shapes = new List<Shape>
        {
            new Circle(2.0),
            new Rectangle(3.0, 4.0),
            new Circle(1.5),
        };

        List<double> areas = new List<double>(shapes.Count);

        Console.WriteLine("Shape areas:");
        foreach (Shape shape in shapes)
        {
            double value = shape.Area();
            areas.Add(value);
            Console.WriteLine($"- {shape.Name()}: {value}");
        }

        double totalArea = 0.0;
        double minArea = areas[0];
        double maxArea = areas[0];

        foreach (double area in areas)
        {
            totalArea += area;
            if (area < minArea)
            {
                minArea = area;
            }

            if (area > maxArea)
            {
                maxArea = area;
            }
        }

        Console.WriteLine($"Total area: {totalArea}");
        Console.WriteLine($"Minimum area: {minArea}");
        Console.WriteLine($"Maximum area: {maxArea}");

        PrintList(new List<int> { 1, 2, 3, 4 }, "Sample counts");
        PrintList(areas, "Computed areas");
    }
}
